src/vcpkg/base/downloads.cpp

```diff
@@ -471,102 +504,163 @@ namespace vcpkg::Downloads
-                            return url;
+                            return true;
                         }
                     }
-                    continue;
+                    else
+                    {
+                        return false;
+                    }
                 }
             }
+        }
 #endif
-            System::Command cmd;
-            cmd.string_arg("curl")
-                .string_arg("--fail")
-                .string_arg("-L")
-                .string_arg(url)
-                .string_arg("--create-dirs")
-                .string_arg("--output")
-                .path_arg(download_path_part_path);
-            const auto out = System::cmd_execute_and_capture_output(cmd);
-            if (out.exit_code != 0)
-            {
-                Strings::append(errors, url, ": ", out.output, '\n');
-                continue;
-            }
+        System::Command cmd;
+        cmd.string_arg("curl")
+            .string_arg("--fail")
+            .string_arg("-L")
+            .string_arg(url)
+            .string_arg("--create-dirs")
+            .string_arg("--output")
+            .path_arg(download_path_part_path);
+        for (auto&& header : headers)
+        {
+            cmd.string_arg("-H").string_arg(header);
+        }
+        const auto out = System::cmd_execute_and_capture_output(cmd);
+        if (out.exit_code != 0)
+        {
+            Strings::append(errors, url, ": ", out.output, '\n');
+            return false;
+        }
 
-            auto maybe_error = try_verify_downloaded_file_hash(fs, url, download_path_part_path, sha512);
-            if (auto err = maybe_error.get())
-            {
-                Strings::append(errors, *err);
-            }
-            else
-            {
-                fs.rename(download_path_part_path, download_path, VCPKG_LINE_INFO);
-                return url;
-            }
+        auto maybe_error = try_verify_downloaded_file_hash(fs, url, download_path_part_path, sha512);
+        if (auto err = maybe_error.get())
+        {
+            Strings::append(errors, *err);
+            return false;
+        }
+        else
+        {
+            fs.rename(download_path_part_path, download_path, VCPKG_LINE_INFO);
+            return true;
+        }
+    }
+
+    static Optional<const std::string&> try_download_files(vcpkg::Files::Filesystem& fs,
+                                                           View<std::string> urls,
+                                                           View<std::string> headers,
+                                                           const fs::path& download_path,
+                                                           const std::string& sha512,
+                                                           std::string& errors)
+    {
+        for (auto&& url : urls)
+        {
+            if (try_download_file(fs, url, headers, download_path, sha512, errors)) return url;
```

Comment:
```suggestion
            if (try_download_file(fs, url, headers, download_path, sha512, errors)) { return url; }
```

Blame Apple's goto fail for this nitpick